src/esp/gfx/CubeMap.cpp

```diff
@@ -0,0 +1,399 @@
+  }
+
+  mapForDraw();
+
+  frameBuffer_.clearDepth(1.0f).clearColor(1, Mn::Vector4ui{0});
+
+  CORRADE_INTERNAL_ASSERT(
+      frameBuffer_.checkStatus(Mn::GL::FramebufferTarget::Draw) ==
+      Mn::GL::Framebuffer::Status::Complete);
+}
+
+void CubeMap::mapForDraw() {
+  frameBuffer_.mapForDraw({
+      {Mn::Shaders::Generic3D::ColorOutput, colorAttachment},
+      // TODO:
+      //{Mn::Shaders::Generic3D::ObjectIdOutput, objectIdAttachment}
+  });
+}
+void CubeMap::textureTypeSanityCheck(TextureType type,
+                                     const std::string& functionNameStr) {
+  switch (type) {
+    case TextureType::Color:
+      CORRADE_ASSERT(flags_ & Flag::ColorTexture,
+                     functionNameStr.c_str()
+                         << "instance was not created with color "
+                            "texture output enabled.", );
+      break;
+    case TextureType::Depth:
+      CORRADE_ASSERT(flags_ & Flag::DepthTexture,
+                     functionNameStr.c_str()
+                         << "instance was not created with depth "
+                            "texture output enabled.", );
+      break;
+    default:
+      CORRADE_INTERNAL_ASSERT_UNREACHABLE();
+      break;
+  }
+}
+
+Mn::GL::CubeMapTexture& CubeMap::getTexture(TextureType type) {
+  textureTypeSanityCheck(type, "CubeMap::getTexture():");
+  return *textures_[type];
+}
+
+std::string CubeMap::getTextureTypeFilenameString(TextureType type) {
+  switch (type) {
+    case TextureType::Color:
+      return std::string(".rgba");
+      break;
+    case TextureType::Depth:
+      return std::string(".depth");
+      break;
+    default:
+      CORRADE_INTERNAL_ASSERT_UNREACHABLE();
+      break;
+  }
+}
+Mn::PixelFormat CubeMap::getPixelFormat(TextureType type) {
+  switch (type) {
+    case TextureType::Color:
+      return Mn::PixelFormat::RGBA8Unorm;
+      break;
+    case TextureType::Depth:
+      return Mn::PixelFormat::R32F;
+      break;
+      /*
+      case TextureType::ObjectId:
+      return Mn::PixelFormat::R32UI;
+      */
+    default:
+      CORRADE_INTERNAL_ASSERT_UNREACHABLE();
+      break;
+  }
+}
+
+bool CubeMap::saveTexture(TextureType type,
+                          const std::string& imageFilePrefix) {
+#ifdef MAGNUM_TARGET_WEBGL
+  // because Mn::Image2D image = textures_[type]->image(...)
+  // requires desktop OpenGL
+  return false;
+#else
+  textureTypeSanityCheck(type, "CubeMap::saveTexture():");
+
+  Cr::PluginManager::Manager<Mn::Trade::AbstractImageConverter> manager;
+  Cr::Containers::Pointer<Mn::Trade::AbstractImageConverter> converter;
+  if (!(converter = manager.loadAndInstantiate("AnyImageConverter"))) {
+    return false;
+  }
+
+  std::string coordStrings[6] = {".+X", ".-X", ".+Y", ".-Y", ".+Z", ".-Z"};
+  for (int iFace = 0; iFace < 6; ++iFace) {
+    // TODO: use framebuffer.read() instead somehow?
+    // But the question is it has NOTHING to do with the framebuffer
+    // anymore at this moment.
```

Comment:
I considered your proposal when I wrote it. 
In OpenGL, changing the rendering target (FBO) is quite expensive. 
This is based on my experience and it matches what is described in "Real-time Rendering (4th)". Here is a screenshot.
![Screenshot from 2021-01-08 15-58-26](https://user-images.githubusercontent.com/931093/104076372-bfea9e80-51ca-11eb-885f-a341cfa24f57.png)
The data is as of 2014 though. I did not verify if it is still the case on the GPU in recent 2 years. I can try it later. In this version, I would like to keep it as this (One FBO).

VK is completely different topic.

Reply:
> In OpenGL, changing the rendering target (FBO) is quite expensive.

Yes, that's correct. But what is commonly *even more* expensive (and what the book doesn't mention) is changing attachments to an existing framebuffer, because the driver has to redo all validation instead of switching to a different but already validated target. :) Though of course it's all just words until a benchmark can prove either way.

> VK is completely different topic.

I don't think it is, GL drivers on Mac are built on top of Metal, which has similar constraints about render targets and pipeline state as Vulkan, and Vulkan was designed based on the experience coming from what had to be done in GL and D3D drivers internally. GL can't just magically be faster at the same thing Vulkan is slow at.

Reply:
You analysis sounds quite reasonable. When I said "VK is completely different topic". I did not mean GL can be faster 
 on something where VK is slow. My point is that experience gained on VK is not necessarily true on GL. For example, here, when being implemented with VK, the method with "6 FBOs" would be faster than the one with "single FBO". It is likely the same case on GL, but I will have to benchmark it to verify.

Reply:
Actually, scratch that. I got a new idea that I think could work much better -- attach all six cubemap faces as six separate color attachments at once and then only call `mapForDraw()` where you draw to a different attachment every time.

Reply:
>attach all six cubemap faces as six separate color attachments at once

Haha! Nice! Thank you. Why did not I think of that? :)